keanu-project/src/test/java/io/improbable/keanu/util/ProgressBarTest.java

```diff
@@ -0,0 +1,95 @@
+package io.improbable.keanu.util;
+
+import static org.junit.Assert.assertEquals;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.ArgumentMatchers.anyLong;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.PrintStream;
+import java.io.UnsupportedEncodingException;
+import java.nio.charset.StandardCharsets;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.atomic.AtomicReference;
+
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+
+public class ProgressBarTest {
+
+    private AtomicReference<Runnable> progressUpdateCall;
+    private ProgressBar progressBar;
+    private ByteArrayOutputStream byteArrayOutputStream;
+
+    @Before
+    public void setup() throws UnsupportedEncodingException {
+
+        byteArrayOutputStream = new ByteArrayOutputStream();
+        PrintStream printStream = new PrintStream(byteArrayOutputStream, true, "UTF-8");
+        ScheduledExecutorService scheduler = mock(ScheduledExecutorService.class);
+
+        progressUpdateCall = new AtomicReference<>(null);
+
+        when(scheduler.scheduleAtFixedRate(any(), anyLong(), anyLong(), any()))
+            .thenAnswer(invocation -> {
+                progressUpdateCall.set(invocation.getArgument(0));
+                return null;
+            });
+
+        progressBar = new ProgressBar(printStream, scheduler);
+    }
+
+    @After
+    public void cleanup() throws IOException {
+        byteArrayOutputStream.close();
+    }
+
+    private void convertCrToNewLine(byte[] outputBytes) {
+        for (int i = 0; i < outputBytes.length; i++) {
+            if (outputBytes[i] == '\r') {
+                outputBytes[i] = '\n';
+            }
+        }
+    }
+
+    private String getResultWithNewLinesInsteadOfCR() {
+        byte[] outputBytes = byteArrayOutputStream.toByteArray();
+        convertCrToNewLine(outputBytes);
+        return new String(outputBytes, StandardCharsets.UTF_8);
+    }
+
+    @Test
+    public void doesPrintToStreamWhenEnabled() {
+        ProgressBar.enable();
+
+        progressBar.progress();
+        progressUpdateCall.get().run();
+        progressUpdateCall.get().run();
+        progressBar.progress();
+        progressUpdateCall.get().run();
+        progressBar.finish();
+
+        String result = getResultWithNewLinesInsteadOfCR();
+
+        assertEquals(5, result.split("\n").length);
```

Comment:
Yeah I thought about that but didn't do it in the end because the real requirement is that it just prints a line of *something*. If that's the requirement then the assertions end up just asserting that there are X number of lines printed. I didn't want to change the formatting of the line printed minimally and break a test. Maybe there should also be an assertion that every new line is different form the previous?